cleanup/cleanup.go

```diff
@@ -0,0 +1,147 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cleanup
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"google.golang.org/grpc"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+	"github.com/grpc/test-infra/config"
+	pb "github.com/grpc/test-infra/proto/grpc/testing"
+	"github.com/grpc/test-infra/status"
+)
+
+// Agent cleanup unwanted processes.
+type Agent struct {
+	client.Client
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	Timeout time.Duration
+}
+
+// +kubebuilder:rbac:groups=e2etest.grpc.io,resources=loadtests,verbs=get;list;watch
+// +kubebuilder:rbac:groups=e2etest.grpc.io,resources=loadtests/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list
+// +kubebuilder:rbac:groups="",resources=pods/status,verbs=get
+
+// Reconcile attempts to check status of workers of the triggering LoadTest, if
+// a terminated LoadTest has workers still running, reconcile will send quit RPC
+// to stop the workers.
+func (a *Agent) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var ctx context.Context
+	var cancel context.CancelFunc
+	var err error
+
+	// A timeout could be set on cleanup agent as a time limit for each reconcile
+	// round, if not set the default value for each reconciliation is 2 mins.
+	if a.Timeout == 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), 2*time.Minute)
+	} else {
+		ctx, cancel = context.WithTimeout(context.Background(), a.Timeout)
+	}
+	defer cancel()
+
+	log := a.Log.WithValues("loadtest", req.NamespacedName)
+
+	// Fetch the LoadTest that triggers the event.
+	loadtest := new(grpcv1.LoadTest)
+	if err = a.Get(ctx, req.NamespacedName, loadtest); err != nil {
+		log.Error(err, "failed to get LoadTest")
+		// do not requeue, the load test may have been deleted
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If the triggering LoadTest is not yet terminated do nothing
+	if !loadtest.Status.State.IsTerminated() {
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch all the pods live on the cluster.
+	pods := new(corev1.PodList)
+	if err = a.List(ctx, pods, client.InNamespace(req.Namespace)); err != nil {
+		log.Error(err, "failed to list pods")
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Reuse existing logic to find the pods related to triggering LoadTest.
+	ownedPods := status.PodsForLoadTest(loadtest, pods.Items)
+
+	q := quitClient{}
+	quitWorkers(ctx, &q, ownedPods, log)
+
+	return ctrl.Result{}, nil
+}
+
+type quit interface {
+	callQuitter(context.Context, *corev1.Pod, logr.Logger)
```

Comment:
My bad. I did not explain it well. I meant that the -er suffix was added to the name of the interface, not the method. For example: [io.Reader](https://golang.org/pkg/io/#Reader) and [io.Writer](https://golang.org/pkg/io/#Writer) from the standard library. So, I would expect the function to be `callQuit(...)` and the interface to be named `CallQuitter`. If you want to use the word "quit" alone, I'd make the interface `Quitter` and the function `quit(...)`.

Reply:
Ah, thank you for the explanation!! :)